src/v/cloud_topics/housekeeper/housekeeper.cc

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/housekeeper/housekeeper.h"
+
+#include "cloud_topics/level_one/metastore/metastore.h"
+#include "cloud_topics/logger.h"
+#include "model/timestamp.h"
+#include "random/simple_time_jitter.h"
+#include "ssx/future-util.h"
+
+#include <chrono>
+#include <exception>
+
+namespace cloud_topics {
+
+housekeeper::housekeeper(
+  model::topic_id_partition tidp,
+  l0_metadata_storage* l0_metastore,
+  l1::metastore* l1_metastore,
+  retention_configuration* config,
+  config::binding<std::chrono::milliseconds> loop_interval)
+  : _tidp(tidp)
+  , _l0_metastore(l0_metastore)
+  , _l1_metastore(l1_metastore)
+  , _config(config)
+  , _loop_interval(std::move(loop_interval)) {}
+
+ss::future<> housekeeper::start() {
+    _gate = {};
+    _as = {};
+    ssx::repeat_until_gate_closed_or_aborted(
+      _gate, _as, [this]() { return do_loop(); });
+    return ss::now();
+}
+
+ss::future<> housekeeper::stop() { co_await _gate.close(); }
+
+ss::future<> housekeeper::do_housekeeping() {
+    if (auto retention_bytes = _config->retention_bytes(_tidp)) {
+        co_await do_bytes_retention(*retention_bytes);
+    }
+    if (auto retention_duration = _config->retention_duration(_tidp)) {
+        co_await do_time_retention(*retention_duration);
+    }
```

Comment:
should we factor out `_l0_metastore->set_start_offset(_tidp, offset);` and set the max once? pretty sure whichever removes the most takes precedence.